drivers/gpio/gpio_keys_zephyr.c

```diff
@@ -0,0 +1,242 @@
+			retval = gpio_remove_callback(gpio_spec->port,
+						      &data->keys[i].cb_data.gpio_cb);
+			memset(&data->keys[i].cb_data, 0, sizeof(struct gpio_keys_callback));
+		}
+		LOG_DBG("disable interrupt [0x%x, %d], rv=%d", gpio_spec->port, gpio_spec->pin,
+			retval);
+	}
+
+	return retval;
+}
+
+static int gpio_keys_get_gpio_port_raw(const struct device *gpio_dev, gpio_port_value_t *value)
+{
+	return z_impl_gpio_port_get_raw(gpio_dev, value);
+}
+
+static int gpio_keys_get_gpio_port_logical(const struct device *gpio_dev, gpio_port_value_t *value)
+{
+	const struct gpio_driver_data *const data = gpio_dev->data;
+
+	int ret = gpio_keys_get_gpio_port_raw(gpio_dev, value);
+
+	if (ret == 0) {
+		*value ^= data->invert;
+	}
+
+	return ret;
+}
+
+static int get_pin(const struct device *dev, uint32_t idx,
+		   int (*gpio_port_get_fn)(const struct device *, gpio_port_value_t *))
+{
+	const struct gpio_keys_config *cfg = dev->config;
+	const struct device *gpio_dev = cfg[idx].spec.port;
+	const struct gpio_driver_config *gpio_cfg = gpio_dev->config;
+	int ret;
+	gpio_port_value_t value;
+
+	__ASSERT((gpio_cfg->port_pin_mask & (gpio_port_pins_t)BIT(cfg[idx].spec.pin)) != 0U,
+		 "Unsupported pin");
+
+	ret = gpio_port_get_fn(gpio_dev, &value);
+
+	if (ret == 0) {
+		ret = (value & (gpio_port_pins_t)BIT(cfg[idx].spec.pin)) != 0 ? 1 : 0;
+	}
+
+	if (ret < 0) {
+		LOG_ERR("Cannot read %s, ret=%d", dev->name, ret);
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static int gpio_keys_zephyr_get_pin(const struct device *dev, uint32_t idx)
+{
+	return get_pin(dev, idx, gpio_keys_get_gpio_port_logical);
+}
+
+static int gpio_keys_zephyr_get_pin_raw(const struct device *dev, uint32_t idx)
+{
+	return get_pin(dev, idx, gpio_keys_get_gpio_port_raw);
+}
+
+static int gpio_keys_init(const struct device *dev)
+{
+	struct gpio_keys_data *data = dev->data;
+
+	for (int i = 0; i < data->num_keys; i++) {
+		data->keys[i].dev = dev;
+	}
+
+	return 0;
+}
+
+static const struct gpio_keys_api gpio_keys_zephyr_api = {
+	.enable_interrupt = gpio_keys_zephyr_enable_interrupt,
+	.disable_interrupt = gpio_keys_zephyr_disable_interrupt,
+	.get_pin = gpio_keys_zephyr_get_pin,
+	.get_pin_raw = gpio_keys_zephyr_get_pin_raw,
+};
+
+#define GPIO_KEYS_CFG_DEF(i)                                                                       \
+	{                                                                                          \
+		.spec = GPIO_DT_SPEC_GET(i, gpios),                                                \
+		.debounce_interval_ms = DT_PROP(i, debounce_interval_ms),                          \
+		.zephyr_code = DT_PROP(i, zephyr_code),                                            \
+	}
+#define GPIO_KEYS_DATA_DEF(i)                                                                      \
+	{                                                                                          \
+		.dev = NULL, .pin_state = -1,                                                      \
```

Comment:
nit: use one entry per line, this looks somewhat incoherent

Reply:
I like one per line as well, clang formatting keeps undoing this.  Let me fix this and see if it passes zephyr compliance checks.